catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergS3SignParams.java

```diff
@@ -0,0 +1,193 @@
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import jakarta.ws.rs.core.Response.Status;
+import java.net.URI;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import org.immutables.value.Value;
+import org.projectnessie.api.v2.params.ParsedReference;
+import org.projectnessie.catalog.files.api.RequestSigner;
+import org.projectnessie.catalog.files.api.SigningRequest;
+import org.projectnessie.catalog.files.api.SigningResponse;
+import org.projectnessie.catalog.files.s3.S3BucketOptions;
+import org.projectnessie.catalog.files.s3.S3Utils;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergException;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergS3SignRequest;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergS3SignResponse;
+import org.projectnessie.catalog.model.snapshot.NessieEntitySnapshot;
+import org.projectnessie.catalog.service.api.CatalogService;
+import org.projectnessie.catalog.service.api.SnapshotReqParams;
+import org.projectnessie.catalog.service.api.SnapshotResponse;
+import org.projectnessie.error.NessieContentNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.IcebergView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+abstract class IcebergS3SignParams {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IcebergS3SignParams.class);
+
+  abstract IcebergS3SignRequest request();
+
+  abstract ParsedReference ref();
+
+  abstract ContentKey key();
+
+  abstract String baseLocation();
+
+  abstract CatalogService catalogService();
+
+  abstract RequestSigner signer();
+
+  abstract S3BucketOptions s3options();
+
+  @Value.Lazy
+  String requestedS3Uri() {
+    return S3Utils.asS3Location(request().uri());
+  }
+
+  @Value.Lazy
+  boolean write() {
+    return request().method().equalsIgnoreCase("PUT")
+        || request().method().equalsIgnoreCase("POST")
+        || request().method().equalsIgnoreCase("DELETE")
+        || request().method().equalsIgnoreCase("PATCH");
+  }
+
+  Uni<IcebergS3SignResponse> verifyAndSign() {
+    return fetchSnapshot()
+        .call(this::checkForbiddenLocations)
+        .onItem()
+        .transformToMulti(this::collectAllowedLocations)
+        .filter(requestedS3Uri()::startsWith)
+        .toUni()
+        .onItem()
+        .ifNull()
+        .failWith(this::unauthorized)
+        .replaceWith(request().uri())
+        .map(this::sign);
+  }
+
+  private Uni<SnapshotResponse> fetchSnapshot() {
+    try {
+      CompletionStage<SnapshotResponse> stage =
+          catalogService()
+              .retrieveSnapshot(
+                  SnapshotReqParams.forSnapshotHttpReq(ref(), "iceberg", null), key(), null);
```

Comment:
Mind adding a TODO here to let `CatalogServiceImpl.retrieveSnapshot()` get a `boolean forWrite` to perform access checks for writes early? #8768 would allow that.